File: Python/0106-m-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(postorder) == 0:
            return None
        if len(postorder) == 1:
            return TreeNode(postorder[-1])
        root = TreeNode(postorder[-1])
        index = inorder.index(root.val)
        root.left = self.buildTree(inorder[0: index], postorder[0: index])
        root.right = self.buildTree(
            inorder[index + 1: len(inorder)], postorder[index: -1])
        return root

File: leetcode/C++/0106-m-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return buildTreeHelper(inorder, 0, inorder.size(), postorder, 0, postorder.size());
    }
    
    TreeNode* buildTreeHelper(vector<int>& inorder, int si, int ei, vector<int>& postorder, int sp, int ep) {
        if (sp == ep) return nullptr;
        TreeNode* root = new TreeNode(postorder[ep - 1]);
        int dis = find(inorder.begin()+si, inorder.begin() + ei, root->val) - inorder.begin() - si;
        root->left = buildTreeHelper(inorder, si, si+dis, postorder, sp, sp+dis);
        root->right = buildTreeHelper(inorder, si+dis+1, ei, postorder, sp+dis, ep-1);
        return root;
    }
};